packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,434 @@
+
+/**
+ * A tree which can be traversed and edited.
+ *
+ * When iterating, only visits non tempt fields.
+ * To discover empty fields, inspect the schema using {@link typeSymbol}.
+ *
+ * TODO: support editing.
+ * TODO: `extends Iterable<EditableField>`
+ * TODO: use proxies for array fields not just raw arrays (will be needed for laziness and editing).
+ * TODO: provide non-schema impacted APIs for getting fields and nodes without unwrapping
+ * (useful for generic code, and when references to these actual fields and nodes are required,
+ * for example creating anchors and editing).
+ */
+export interface EditableTree {
+    /**
+     * The type of this node.
+     * If this node is is well-formed, it must follow this schema.
+     */
+    readonly [type]: TreeSchema;
+
+    /**
+     * A function to get the type of a non-sequence child node.
+    */
+    readonly [getTypeName]: (key: string) => TreeSchemaIdentifier;
+
+    /**
+     * Value stored on this node.
+     */
+    readonly [value]: Value;
+
+    readonly [proxySymbol]: object;
+
+    /**
+     * Fields of this node, indexed by their field keys (as strings).
+     *
+     * This API exposes content in a way depending on the {@link Multiplicity} of the {@link FieldKind}.
+     * Sequences (including empty ones) are always exposed as arrays,
+     * and everything else is either a single EditableTree or undefined depending on if it's empty.
+     *
+     * TODO:
+     * This approach to field lookup can result in collisions between global and local keys,
+     * particularly with "extra" fields.
+     * A mechanism for disambiguating this should be added,
+     * likely involving an alternative mechanism for looking up global fields via symbols.
+     */
+    readonly [key: string]: UnwrappedEditableField;
+}
+
+/**
+ * EditableTree,
+ * but with for a type that `isPrimitive` unwrapped into just the value if that value is a {@link PrimitiveValue}.
+ */
+export type EditableTreeOrPrimitive = EditableTree | PrimitiveValue;
+
+/**
+ * EditableTree, but with these cases of unwrapping:
+ * - primitives are unwrapped. See {@link EditableTreeOrPrimitive}.
+ * - nodes with PrimaryField are unwrapped to just the primaryField. See `getPrimaryField`.
+ * - fields are unwrapped based on their schema's multiplicity. See {@link UnwrappedEditableField}.
+ *
+ * TODO:
+ * EditableTree should provide easy access to children in a way thats guaranteed
+ * not to do this unwrapping for cases which need to refer to the actual nodes.
+ * This may include cases like creating anchors and/or editing.
+ */
+export type UnwrappedEditableTree = EditableTreeOrPrimitive | readonly UnwrappedEditableTree[];
+
+/**
+ * A field of an {@link EditableTree}.
+ */
+export type EditableField = readonly [FieldSchema, readonly EditableTree[]];
+
+/**
+ * Unwraps fields with non-sequence multiplicities.
+ */
+export type UnwrappedEditableField = UnwrappedEditableTree | undefined | readonly UnwrappedEditableTree[];
```

Comment:
Is the `readonly UnwrappedEditableTree[]` case for array nodes? If so, it would be good to indicate that in comment as the presence of `readonly UnwrappedEditableTree[]` is surprising here.

Reply:
That case is for sequence fields. I have added a suggestion above for clarifying this comment.